### ISSUE TITLE
Code for Experiments: Unsupervised mesh prediction

### ISSUE THREAD

**Issue by vikashranjan** [facebookresearch/pytorch3d#306]
Is the code for Experiments with Sphere FC, Sphere GCN, and Voxel GCN mentioned in the paper "Accelerating 3D Deep Learning with PyTorch3D" available?

As I am new to 3d deep learning I would like to go through the code to understand Unsupervised mesh prediction.

**Comment by gkioxari** [comment]
Our plan is to have this soon! In the mean time we have a tutorial on fitting a mesh from image views using differentiable rendering that could be helpful and that don't deviate much from the learning setting (mesh verts are predicted via a neural network instead of being treated as a variable tensor):

https://github.com/facebookresearch/pytorch3d/blob/master/docs/tutorials/fit_textured_mesh.ipynb

**Comment by vikashranjan** [comment]
Thanks, looking forward to the code release!

**Comment by sjcv** [comment]
Hi @gkioxari , just wanted to check on the timeline for this. Thanks again for providing it!

**Comment by nikhilaravi** [comment]
@sjcv we are planning to open sourcing this as a part of a new `projects` section in PyTorch3D. We will comment on this task when the code is available.

**Comment by JobAtom** [comment]
> @sjcv we are planning to open sourcing this as a part of a new `projects` section in PyTorch3D. We will comment on this task when the code is available.

In SphereGCN, 3xGraphConv(512 + 3, 512) + ReLU layer, I am wondering what is 512 + 3 dimension input? Is the 3 means the x, y, z position of input vertex? Thank you.

**Comment by gkioxari** [comment]
Yes!

**Comment by stephentandjiria** [comment]
Hi, I'm super interested in unsupervised 3D prediction from a 2D image.
However, I am confused as to how to train the following NN
![image](https://user-images.githubusercontent.com/75295743/117319783-d3c2ba80-aeb5-11eb-9edf-e1c66e498390.png)

in a way that minimizes all loss functions (rgb, silhouette, edge, normal, laplacian) inside the loop in [fit_textured_mesh.ipynb](https://github.com/facebookresearch/pytorch3d/blob/master/docs/tutorials/fit_textured_mesh.ipynb)

I noticed that the only two ways in comparing prediction to ground truth are silhouette and RGB.
```
    for j in np.random.permutation(num_views).tolist()[:num_views_per_iteration]:
        images_predicted = renderer_textured(new_src_mesh, cameras=target_cameras[j], lights=lights)

        # Squared L2 distance between the predicted silhouette and the target 
        # silhouette from our dataset
        predicted_silhouette = images_predicted[..., 3]
        loss_silhouette = ((predicted_silhouette - target_silhouette[j]) ** 2).mean()
        loss["silhouette"] += loss_silhouette / num_views_per_iteration
        
        # Squared L2 distance between the predicted RGB image and the target 
        # image from our dataset
        predicted_rgb = images_predicted[..., :3]
        loss_rgb = ((predicted_rgb - target_rgb[j]) ** 2).mean()
        loss["rgb"] += loss_rgb / num_views_per_iteration
```
 
but I am confused as to how to pull out the silhouette and RGB as outputs from the NN.

or maybe I should pull out the NN verts output and put it into src_mesh? replacing the following line:
`src_mesh = ico_sphere(4, device)`
and then just continue the deformation loop?

**Comment by stephentandjiria** [comment]
Focusing further from my unstructured previous question,

I found that out of the 5 loss functions, edge/normal/laplacian take care of itself (i.e. implement just like fit_textured_mesh.ipynb without any input from NN). So we need to focus on **loss_silhouette** and **loss_rgb**.

Now:
- **loss_silhouette** is from **predicted_silhouette** and **target_silhouette**
- **loss_rgb** is from **predicted_rgb** and **target_rgb**

### 1. predicted_silhouette
```
# new_src_mesh is Sphere GCN NN output verts(?) (B x 642 x 3)
new_src_mesh.textures = TexturesVertex(verts_features=rgb_textures_from_nn) # (B x 642 x 3 from Sphere GCN NN output texture)
images_predicted = renderer_textured(new_src_mesh, cameras=target_cameras[j], lights=lights)
predicted_silhouette = images_predicted[..., 3]
```
- but why do we use **renderer_textured** instead of **renderer_silhouette** where the output is supposed to be a silhouette?
- is **new_src_mesh** really the **Sphere GCN NN verts output**?

### 2. target_silhouette
```
mesh = load_objs_as_meshes([obj_filename], device=device)
meshes = mesh.extend(num_views)
silhouette_images = renderer_silhouette(meshes, cameras=cameras, lights=lights)
target_silhouette = [silhouette_images[i, ..., 3] for i in range(num_views)]
```
- if we want to get **silhouette_images from 2D image instead of a 3D mesh**, is there a **"renderer_silhouette" just for 2D**?

### 3. predicted_rgb
```
# new_src_mesh is Sphere GCN NN output verts(?) (B x 642 x 3)
images_predicted = renderer_textured(new_src_mesh, cameras=target_cameras[j], lights=lights)
predicted_rgb = images_predicted[..., :3]
```
- already clear

### 4. target_rgb
```
mesh = load_objs_as_meshes([obj_filename], device=device)
meshes = mesh.extend(num_views)
silhouette_images = renderer_silhouette(meshes, cameras=cameras, lights=lights)
target_images = renderer(meshes, cameras=cameras, lights=lights)
target_rgb = [target_images[i, ..., :3] for i in range(num_views)]
```
- same with (2), if we want to get **silhouette_images from 2D image instead of a 3D mesh**, is there a **"renderer_silhouette" just for 2D**?

### 5. Optimizer
Would it be something like this instead?
`optimizer = torch.optim.SGD(sphere_gcn_model.parameters(), lr=1.0, momentum=0.9)`

Sorry if my questions sound too basic as I am very new in deep learning. Thank you!

**Comment by stephentandjiria** [comment]
Solved it.

To anyone wondering how to create the model in the paper, these are useful for me:
- https://pytorch3d.readthedocs.io/en/latest/_modules/pytorch3d/ops/cubify.html
- https://github.com/facebookresearch/meshrcnn/blob/master/shapenet/modeling/heads/mesh_head.py#L112
- https://pytorch3d.readthedocs.io/en/latest/modules/ops.html#pytorch3d.ops.vert_align

and just follow this, especially the padded/packed arrangement (sourced from PyTorch team SIGGRAPH presentation):
![image](https://user-images.githubusercontent.com/75295743/117575411-d585bb80-b10b-11eb-80ad-d1c3d485b0b8.png)

It is not specifically mentioned in the paper, but I suspect the last layer with **tanh activation is verts**, and the one with **sigmoid activation is for texture**. Although, I am still trying to find out which one is for faces.

**Comment by sabraha2** [comment]
Also interested in this! :)

**Comment by moyans** [comment]
Any updates here?
